actionpack/lib/action_dispatch/routing/route_set.rb

```diff
@@ -822,10 +822,12 @@ def url_for(options, route_name = nil, url_strategy = UNKNOWN, method_name = nil
         path = route_with_params.path(method_name)
         params = route_with_params.params
 
-        if options[:params].is_a?(Hash)
-          params.merge! options[:params]
-        elsif options.key? :params
-          params[:params] = options[:params]
+        if options.key? :params
```

Comment:
I don't understand why you made a nest conditional. You already use the safe operator for checking `options[:params]&.respond_to?(:to_hash)`

Reply:
```suggestion
        if options[:params]&.respond_to?(:to_hash)
          params.merge! options[:params]
        elsif options.key? :params
          params[:params] = options[:params]
        end
```